cfgrammar/src/lib/yacc/reporting.rs

```diff
@@ -7,6 +11,38 @@ pub trait ErrorReport {
     fn grammar_error(&mut self, err: YaccGrammarError);
     /// Gives ownership of a warning to self.
     fn grammar_warning(&mut self, err: YaccGrammarWarning);
+
+    /// Default implementation does nothing. If you override this
```

Comment:
Do we need a function for this -- can't the user do it with a `drop` method?

[The idea of a `Stage` below is interesting. I'm mildly nervous about it, though, because it does somewhat bake in some aspects of the implementation that aren't fundamental. I also wonder if a user who cares about these distinctions can already infer it from the error kinds they see?]

Reply:
We actually can't usefully implement it from `Drop`, because of the following sequence:

```
let foo = DedupReport(ErrorReport::new());
gb.build(); // Assume errors are piling up in `DedupReport` but don't call finish();
assert_eq!(foo.format_errors(...).collect<Vec<_>>(), vec![]);
drop(foo); // Too late, it's the call to `foo` that depends upon the call to finish(), since errors are buffering in `DedupReport`. But `format_errors` delegates to `child_report`. 
```

If were to then say call finish within `format_errors`, we just need to be careful that we never feed an error multiple times in the case where `finish` was already called (i.e. drain the errors as it does, so in theory we could maybe drain the). 

I *think* perhaps we could drop `finish` if we just drain to `child` in `format_errors` without a call to `finish`.
The consequences of that is that if you call `format_errors` while errors are still arriving you could be getting duplicates.

I think that is the best way I can explain it.

Reply:
I updated the example code snippet in the comment above to hopefully better reflect the problem.

Reply:
One thing though we can do, is call this function from `ReportHandler::drop`, potentially even giving it an sealed argument so it can't be called outside of report handler, but that is exactly the same point I called it from `build()` chain above.